# The 4th Down Equation: Analyzing the Interaction Between Offensive EPA and Defensive Pass-Prevention Tendencies

**Team Members:** Nathan Spear, Hayden Kellington, Shivam Sharma

**Date:** 02/11/206

---

## Research Question

How does going for it on 4th down affect Expected Points Added (EPA) on the play, and does this effect differ based on a defense’s pass-prevention tendencies, measured by opponents’ average depth of target (aDOT) allowed and EPA allowed on deep passing plays?

**Expected Outcomes:**
- Quantifying the "Go for it" Impact: We aim to demonstrate that going for it on 4th down generally results in a positive Expected Points Added (EPA) compared to more conservative play-calling, particularly in short-yardage situations.


- Defensive Shell Sensitivity: We hope to discover if offensive success on 4th down is significantly hindered by defenses with low average depth of target (aDOT) allowed—which suggests a "bend-but-don't-break" or deep-pass-prevention style.


- The "Deep-Pass EPA" Correlation: We want to test the hypothesis that teams facing a defense with a high Deep-Pass EPA allowed will see a higher marginal benefit (EPA gain) when going for it on 4th down passing plays.


- Interaction Effects: We aim to demonstrate that the decision to go for it is not "one size fits all" but is statistically dependent on defensive tendencies, showing that an offense's EPA outcome on 4th down varies predictably when interacting with a defense's historical pass-prevention metrics.


- Data-Driven Decision Making: Ultimately, we hope to provide a framework that moves beyond "offensive confidence" to show that specific pbp-derived defensive metrics (like aDOT and deep-pass EPA) should be primary factors in a team’s 4th-down strategy.

---

## Data Source

**Dataset Name:** nflverse (specifically nflfastR play-by-play data)

**Link:** https://github.com/nflverse/nflfastR-data

**Description:** 
A comprehensive dataset containing play-by-play information for every NFL game. It includes advanced metrics like Expected Points Added (EPA), Win Probability (WP), and specific play details such as air yards, pass location, and personnel.

- Number of observations: Approximately 45,000 to 50,000 plays per regular season. For your specific study on 4th downs, this will be filtered down to roughly 4,000–5,000 observations depending on the years selected.


- Number of features: 370+ variables per play, covering everything from game state (time, score, yard line) to advanced efficiency metrics.

Key variables:


- epa: Expected Points Added on the play (Target Variable).


- fourth_down_attempt: Binary indicator (1 if the team went for it, 0 otherwise).


- air_yards: Distance the ball traveled in the air, used to calculate aDOT.


- pass_length / pass_location: Used to identify deep passing plays.


- defteam: The defensive team, used to aggregate defensive tendency metrics.


- yardline_100: Distance to the opponent's end zone.

- Time period covered: The nflverse repository contains data from 1999 to the 2024 season. Based on your project guidance, you will likely focus on the most recent 2024–2025 data.


- Data collection method: Scraped and cleaned play-by-play data sourced directly from the NFL’s Game Statistics and Information System (GSIS).


**Citation:** 
Carl, S., & Baldwin, B. (2024). nflfastR: Functions to Efficiently Access NFL Play by Play Data. https://www.nflverse.com/.

---

## Setup and Imports

In [2]:
# Standard imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import scipy.stats as stats

!pip install nfl_data_py
import nfl_data_py as nfl

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
plt.style.use('seaborn-v0_8-darkgrid')
sns.set_palette('husl')

# For reproducibility
np.random.seed(42)

print("Imports successful!")

Defaulting to user installation because normal site-packages is not writeable
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 24.5 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 60.1 MB/s eta 0:00:0000:010:01m
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 52.3 MB/s eta 0:00:0000:01
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
geopandas 1.1.2 requires pandas>=2.0.0, but you have pandas 1.5.3 which is incompatible.
libpysal 4.14.1 requires pandas>=2.1.0, but you have pandas 1.5.3 which is incompatible.
dask-expr 1.0.11 requires pandas>=2, but you have pandas 1.5.3 which is incompatible.
Imports successful!


---

## Data Loading

**TODO:** Load your dataset and perform initial inspection

In [6]:
df = nfl.import_pbp_data([2024, 2025])
df_4th = df[df['down'] == 4].copy()
print(f"Successfully loaded {len(df_4th)} fourth-down plays!")

# Display basic information
if df is not None:
    print(f"Dataset shape: {df_4th.shape}")
    print(f"\nFirst few rows:")
    display(df_4th.head())

2024 done.
2025 done.
Downcasting floats.
Successfully loaded 8569 fourth-down plays!
Dataset shape: (8569, 397)

First few rows:


play_id          game_id old_game_id_x home_team away_team season_type  \
31     823.0  2024_01_ARI_BUF    2024090801       BUF       ARI         REG   
43    1120.0  2024_01_ARI_BUF    2024090801       BUF       ARI         REG   
93    2360.0  2024_01_ARI_BUF    2024090801       BUF       ARI         REG   
97    2468.0  2024_01_ARI_BUF    2024090801       BUF       ARI         REG   
103   2632.0  2024_01_ARI_BUF    2024090801       BUF       ARI         REG   

     week posteam posteam_type defteam side_of_field  yardline_100  \
31      1     ARI         away     BUF           BUF          11.0   
43      1     BUF         home     ARI           ARI          19.0   
93      1     ARI         away     BUF           ARI          74.0   
97      1     BUF         home     ARI           BUF          66.0   
103     1     BUF         home     ARI           ARI          12.0   

      game_date  quarter_seconds_remaining  half_seconds_remaining  \
31   2024-09-08                      900.0                   900.0   
43   2024-09-08                      591.0                   591.0   
93   2024-09-08                      483.0                  1383.0   
97   2024-09-08                      388.0                  1288.0   
103  2024-09-08                      204.0                  1104.0   

     game_seconds_remaining game_half  quarter_end  drive   sp  qtr  down  \
31                   2700.0     Half1          0.0    3.0  1.0  2.0   4.0   
43                   2391.0     Half1          0.0    4.0  1.0  2.0   4.0   
93                   1383.0     Half2          0.0    9.0  0.0  3.0   4.0   
97                   1288.0     Half2          0.0   10.0  0.0  3.0   4.0   
103                  1104.0     Half2          0.0   12.0  0.0  3.0   4.0   

     goal_to_go   time   yrdln  ydstogo  ydsnet  \
31            0  15:00  BUF 11      4.0    60.0   
43            1  09:51  ARI 19     19.0    51.0   
93            0  08:03  ARI 26     14.0    -4.0   
97            0  06:28  BUF 34      7.0     3.0   
103           0  03:24  ARI 12      1.0    21.0   

                                                  desc   play_type  \
31   (15:00) 5-M.Prater 29 yard field goal is GOOD,...  field_goal   
43   (9:51) 2-T.Bass 37 yard field goal is GOOD, Ce...  field_goal   
93   (8:03) 12-B.Gillikin punts 50 yards to BUF 24,...        punt   
97   (6:28) 8-S.Martin punts 51 yards to ARI 15, Ce...        punt   
103  (3:24) 17-J.Allen up the middle to ARI 11 for ...         run   

     yards_gained  shotgun  no_huddle  qb_dropback  qb_kneel  qb_spike  \
31            0.0      0.0        0.0          0.0       0.0       0.0   
43            0.0      0.0        0.0          0.0       0.0       0.0   
93            0.0      0.0        0.0          0.0       0.0       0.0   
97            0.0      0.0        0.0          0.0       0.0       0.0   
103           1.0      0.0        0.0          0.0       0.0       0.0   

     qb_scramble pass_length pass_location  air_yards  yards_after_catch  \
31           0.0        None          None        NaN                NaN   
43           0.0        None          None        NaN                NaN   
93           0.0        None          None        NaN                NaN   
97           0.0        None          None        NaN                NaN   
103          0.0        None          None        NaN                NaN   

    run_location run_gap field_goal_result  kick_distance extra_point_result  \
31          None    None              made           29.0               None   
43          None    None              made           37.0               None   
93          None    None              None           50.0               None   
97          None    None              None           51.0               None   
103       middle    None              None            NaN               None   

    two_point_conv_result  home_timeouts_remaining  away_timeouts_remaining  \
31                   None          

In [7]:
# TODO: Examine dataset structure
if df_4th is not None:
    print("Dataset Info:")
    df_4th.info()
    
    print("\n" + "="*50)
    print("Summary Statistics:")
    display(df_4th.describe())
    
    print("\n" + "="*50)
    print("Data Types:")
    display(df_4th.dtypes)

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
Index: 8569 entries, 31 to 98253
Columns: 397 entries, play_id to defense_numbers
dtypes: float32(205), int32(7), int64(1), object(184)
memory usage: 19.1+ MB

Summary Statistics:


play_id         week  yardline_100  quarter_seconds_remaining  \
count  8569.000000  8569.000000   8569.000000                8569.000000   
mean   2217.985352     9.946552     47.471699                 403.379395   
std    1262.482666     5.602199     24.115273                 265.885742   
min     120.000000     1.000000      1.000000                   1.000000   
25%    1107.000000     5.000000     28.000000                 160.000000   
50%    2220.000000    10.000000     50.000000                 386.000000   
75%    3313.000000    15.000000     66.000000                 630.000000   
max    5321.000000    22.000000     99.000000                 900.000000   

       half_seconds_remaining  game_seconds_remaining  quarter_end  \
count             8569.000000             8569.000000       8569.0   
mean               781.801636             1664.471680          0.0   
std                533.047852             1047.310791          0.0   
min                  1.000000                1.000000          0.0   
25%                285.000000              725.000000          0.0   
50%                764.000000             1713.000000          0.0   
75%               1241.000000             2559.000000          0.0   
max               1785.000000             3577.000000          0.0   

             drive           sp          qtr    down   goal_to_go  \
count  8569.000000  8569.000000  8569.000000  8569.0  8569.000000   
mean     11.047497     0.232583     2.603804     4.0     0.059867   
std       6.480571     0.422499     1.129239     0.0     0.237254   
min       1.000000     0.000000     1.000000     4.0     0.000000   
25%       5.000000     0.000000     2.000000     4.0     0.000000   
50%      11.000000     0.000000     3.000000     4.0     0.000000   
75%      16.000000     0.000000     4.000000     4.0     0.000000   
max      29.000000     1.000000     5.000000     4.0     1.000000   

           ydstogo       ydsnet  yards_gained      shotgun    no_huddle  \
count  8569.000000  8569.000000   8541.000000  8569.000000  8569.000000   
mean      7.631462    26.023457      0.962651     0.158828     0.024740   
std       5.743312    25.178387      4.391374     0.365535     0.155337   
min       1.000000   -35.000000    -34.000000     0.000000     0.000000   
25%       3.000000     5.000000      0.000000     0.000000     0.000000   
50%       6.000000    19.000000      0.000000     0.000000     0.000000   
75%      10.000000    45.000000      0.000000     0.000000     0.000000   
max      40.000000    99.000000     55.000000     1.000000     1.000000   

       qb_dropback     qb_kneel  qb_spike  qb_scramble   air_yards  \
count  8549.000000  8569.000000    8569.0  8569.000000  999.000000   
mean      0.134519     0.000583       0.0     0.008052   10.056056   
std       0.341216     0.024150       0.0     0.089376    9.991424   
min       0.000000     0.000000       0.0     0.000000  -11.000000   
25%       0.000000     0.000000       0.0     0.000000    3.000000   
50%       0.000000     0.000000       0.0     0.000000    7.000000   
75%       0.000000     0.000000       0.0     0.000000   14.000000   
max       1.000000     1.000000       0.0     1.000000   49.000000   

       yards_after_catch  kick_distance  home_timeouts_remaining  \
count         566.000000    6248.000000              8569.000000   
mean            3.985866      44.716709                 2.553740   
std             5.841841      10.667015                 0.795331   
min            -2.000000       0.000000                 0.000000   
25%             0.000000      38.000000                 2.000000   
50%             1.500000      46.000000                 3.000000   
75%             6.000000      52.000000                 3.000000   
max            41.000000      84.000000                 3.000000   

       away_timeouts_remaining      timeout  posteam_timeouts_remaining  \
count              8569.000000  8541.000000                 8569.000000   
mea


Data Types:


play_id              float32
game_id               object
old_game_id_x         object
home_team             object
away_team             object
                      ...   
defense_names         object
offense_positions     object
defense_positions     object
offense_numbers       object
defense_numbers       object
Length: 397, dtype: object

**Initial Observations:**

- Noise & Complexity: With 397 columns, most data (like jersey numbers or lateral stats) is irrelevant and needs to be dropped to focus on EPA and defensive metrics.

- Missing Values (Sparsity): Metrics like air_yards are missing in ~60% of rows because they only apply to passes. We have to filter for pass_attempt == 1 before calculating aDOT to avoid skewed results.

- Data Integrity: There are roughly 1,000 rows missing EPA (likely timeouts or penalties) which should be removed. While binary indicators (like fourth_down_converted) are stored as floats, they are mathematically sound for regression.

- Range Validation: EPA values (-12.7 to +8.5) and yard lines (1 to 99) are within expected NFL ranges. The 4th-down conversion rate is currently low (~0.6%) because the data includes all downs; this will normalize once filtered for 4th-down attempts only.

---

## Data Cleaning

**TODO:** Clean and preprocess the data

### Missing Values Analysis

In [ ]:
# TODO: Check for missing values
if df is not None:
    missing = df.isnull().sum()
    missing_pct = (missing / len(df)) * 100
    
    missing_df = pd.DataFrame({
        'Missing Count': missing,
        'Percentage': missing_pct
    }).sort_values('Percentage', ascending=False)
    
    print("Missing Values Summary:")
    display(missing_df[missing_df['Missing Count'] > 0])
    
    # Visualize missing data
    if missing.sum() > 0:
        plt.figure(figsize=(10, 6))
        missing_df[missing_df['Missing Count'] > 0]['Percentage'].plot(kind='barh')
        plt.xlabel('Percentage Missing')
        plt.title('Missing Values by Column')
        plt.tight_layout()
        plt.show()

In [ ]:
# TODO: Handle missing values
# Strategy options:
# 1. Drop rows: df = df.dropna()
# 2. Drop columns: df = df.drop(columns=['col_name'])
# 3. Fill with mean/median: df['col'] = df['col'].fillna(df['col'].mean())
# 4. Fill with mode: df['col'] = df['col'].fillna(df['col'].mode()[0])
# 5. Forward/backward fill: df = df.fillna(method='ffill')

# df_clean = df.copy()
# TODO: Implement your cleaning strategy here

### Duplicate Detection

In [ ]:
# TODO: Check for duplicates
if df is not None:
    duplicates = df.duplicated().sum()
    print(f"Number of duplicate rows: {duplicates}")
    
    if duplicates > 0:
        print("\nDuplicate rows:")
        display(df[df.duplicated(keep=False)])
        
        # TODO: Decide whether to keep or remove duplicates
        # df_clean = df_clean.drop_duplicates()

### Data Type Conversions

In [ ]:
# TODO: Convert data types as needed
# Examples:
# df_clean['date_column'] = pd.to_datetime(df_clean['date_column'])
# df_clean['category_column'] = df_clean['category_column'].astype('category')
# df_clean['numeric_column'] = pd.to_numeric(df_clean['numeric_column'], errors='coerce')

pass

### Outlier Detection

In [ ]:
# TODO: Detect outliers in numeric columns
# Common methods:
# 1. IQR method
# 2. Z-score method
# 3. Visual inspection with box plots

# Example: Box plots for numeric columns
if df is not None:
    numeric_cols = df.select_dtypes(include=[np.number]).columns
    
    if len(numeric_cols) > 0:
        # TODO: Create box plots for numeric columns
        # fig, axes = plt.subplots(len(numeric_cols), 1, figsize=(10, 3*len(numeric_cols)))
        # for i, col in enumerate(numeric_cols):
        #     df.boxplot(column=col, ax=axes[i])
        # plt.tight_layout()
        # plt.show()
        pass

### Feature Engineering (Optional)

In [ ]:
# TODO: Create new features if needed
# Examples:
# - Combine existing features
# - Extract date components (year, month, day of week)
# - Bin continuous variables
# - Encode categorical variables

pass

In [ ]:
# TODO: Save cleaned dataset (optional)
# df_clean.to_csv('data/cleaned_data.csv', index=False)
# print("Cleaned data saved!")

**Cleaning Summary:**

TODO: Document what cleaning steps were performed and why:
- Missing values: [strategy used]
- Duplicates: [action taken]
- Outliers: [how handled]
- Feature engineering: [new features created]

---

## Exploratory Data Analysis

**TODO:** Explore the data to understand patterns, relationships, and distributions

### Univariate Analysis

In [ ]:
# TODO: Analyze distributions of individual variables

# Numeric variables - histograms and density plots
# if df_clean is not None:
#     numeric_cols = df_clean.select_dtypes(include=[np.number]).columns
#     
#     for col in numeric_cols:
#         fig, axes = plt.subplots(1, 2, figsize=(12, 4))
#         
#         df_clean[col].hist(bins=30, ax=axes[0], edgecolor='black')
#         axes[0].set_title(f'Histogram of {col}')
#         axes[0].set_xlabel(col)
#         
#         df_clean[col].plot(kind='density', ax=axes[1])
#         axes[1].set_title(f'Density Plot of {col}')
#         axes[1].set_xlabel(col)
#         
#         plt.tight_layout()
#         plt.show()

pass

In [ ]:
# Categorical variables - bar charts
# if df_clean is not None:
#     categorical_cols = df_clean.select_dtypes(include=['object', 'category']).columns
#     
#     for col in categorical_cols:
#         plt.figure(figsize=(10, 5))
#         df_clean[col].value_counts().plot(kind='bar', edgecolor='black')
#         plt.title(f'Distribution of {col}')
#         plt.xlabel(col)
#         plt.ylabel('Count')
#         plt.xticks(rotation=45)
#         plt.tight_layout()
#         plt.show()

pass

### Bivariate Analysis

In [ ]:
# TODO: Explore relationships between pairs of variables

# Correlation matrix for numeric variables
# if df_clean is not None:
#     numeric_cols = df_clean.select_dtypes(include=[np.number]).columns
#     
#     if len(numeric_cols) > 1:
#         plt.figure(figsize=(10, 8))
#         correlation_matrix = df_clean[numeric_cols].corr()
#         sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', center=0, 
#                     square=True, linewidths=1)
#         plt.title('Correlation Matrix')
#         plt.tight_layout()
#         plt.show()

pass

In [ ]:
# TODO: Scatter plots for key variable pairs
# plt.figure(figsize=(8, 6))
# plt.scatter(df_clean['var1'], df_clean['var2'], alpha=0.5)
# plt.xlabel('Variable 1')
# plt.ylabel('Variable 2')
# plt.title('Relationship between Var1 and Var2')
# plt.show()

pass

In [ ]:
# TODO: Group comparisons
# if df_clean is not None:
#     # Example: Compare numeric variable across categories
#     # df_clean.groupby('category_col')['numeric_col'].describe()
#     
#     # Box plot by group
#     # plt.figure(figsize=(10, 6))
#     # sns.boxplot(x='category_col', y='numeric_col', data=df_clean)
#     # plt.xticks(rotation=45)
#     # plt.tight_layout()
#     # plt.show()
    
    pass

### Multivariate Analysis

In [ ]:
# TODO: Explore relationships among multiple variables

# Pair plot for selected variables
# if df_clean is not None:
#     # Select key columns for pair plot
#     # key_cols = ['col1', 'col2', 'col3', 'target']
#     # sns.pairplot(df_clean[key_cols], hue='target')
#     # plt.show()
    
    pass

**EDA Findings:**

TODO: Summarize key insights from your exploratory analysis:
- What are the main patterns in the data?
- Are there any unexpected findings?
- Which variables seem most relevant to your research question?
- Are there any data quality issues that need addressing?

---

## Modeling and Analysis

**TODO:** Build and evaluate models to answer your research question

### Data Preparation for Modeling

In [ ]:
# TODO: Prepare data for modeling
# from sklearn.model_selection import train_test_split
# from sklearn.preprocessing import StandardScaler, LabelEncoder

# Define features and target
# X = df_clean[['feature1', 'feature2', 'feature3']]
# y = df_clean['target']

# Split data
# X_train, X_test, y_train, y_test = train_test_split(
#     X, y, test_size=0.2, random_state=42
# )

# Scale features (if needed)
# scaler = StandardScaler()
# X_train_scaled = scaler.fit_transform(X_train)
# X_test_scaled = scaler.transform(X_test)

pass

### Model 1: [Model Name]

**TODO:** Describe the model and why you chose it

In [ ]:
# TODO: Train your first model
# from sklearn.linear_model import LogisticRegression
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# model = LogisticRegression(random_state=42)
# model.fit(X_train_scaled, y_train)

pass

In [ ]:
# TODO: Make predictions and evaluate
# y_pred = model.predict(X_test_scaled)

# print("Model Performance:")
# print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
# print("\nClassification Report:")
# print(classification_report(y_test, y_pred))

# Confusion Matrix
# cm = confusion_matrix(y_test, y_pred)
# plt.figure(figsize=(8, 6))
# sns.heatmap(cm, annot=True, fmt='d', cmap='Blues')
# plt.title('Confusion Matrix')
# plt.ylabel('True Label')
# plt.xlabel('Predicted Label')
# plt.show()

pass

### Model 2: [Model Name]

**TODO:** Describe your second model approach

In [ ]:
# TODO: Train and evaluate second model

pass

### Model Comparison

In [ ]:
# TODO: Compare model performance
# Create a comparison table or visualization

# results_df = pd.DataFrame({
#     'Model': ['Model 1', 'Model 2'],
#     'Accuracy': [acc1, acc2],
#     'Precision': [prec1, prec2],
#     'Recall': [rec1, rec2],
#     'F1-Score': [f1_1, f1_2]
# })
# display(results_df)

pass

### Feature Importance (Optional)

In [ ]:
# TODO: Analyze feature importance (if applicable)
# if hasattr(model, 'feature_importances_'):
#     importance_df = pd.DataFrame({
#         'Feature': X.columns,
#         'Importance': model.feature_importances_
#     }).sort_values('Importance', ascending=False)
#     
#     plt.figure(figsize=(10, 6))
#     plt.barh(importance_df['Feature'], importance_df['Importance'])
#     plt.xlabel('Importance')
#     plt.title('Feature Importance')
#     plt.gca().invert_yaxis()
#     plt.tight_layout()
#     plt.show()

pass

**Modeling Results:**

TODO: Summarize your modeling findings:
- Which model performed best and why?
- What are the most important predictors?
- Are there any limitations or concerns with the models?
- Do the results answer your research question?

---

## Conclusions and Future Work

**TODO:** Summarize your project and findings

### Key Findings

TODO: List your main discoveries:
1. 
2. 
3. 

### Limitations

TODO: What are the limitations of this analysis?
- 
- 

### Future Work

TODO: What could be done to extend or improve this analysis?
- 
- 

### Recommendations

TODO: Based on your findings, what actions or decisions do you recommend?
- 
- 

---

## References

TODO: List all data sources, papers, and resources used:

1. 
2. 
3. 